### PR TITLE
somnia bot

### PR DESCRIPTION
**Pull request by Godlander**
## Describe your changes

Add barebone code for new somnia discord bot

## Issue or discord link

https://discord.com/channels/785153694478893126/1220796004442312807

## Testing/validation

image

## Checklist before requesting a review (leave this PR as draft if any part of this list is not done.)

- [x] I have commented my code in hard-to understand areas.
- [ ] I have made corresponding changes to README or wiki.
- [ ] For front-end changes, I have updated the corresponding English translations.
- [x] I have run `yarn run mini-ci` locally to validate format and lint.
- [ ] If I have added a new library or app, I have updated the deployment scripts to ignore changes as needed

### CODE REVIEW

**Review by nguyentvan7 — DISMISSED** [review]

File: apps/somnia/project.json [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+      "options": {
+        "platform": "node",
+        "outputPath": "dist/apps/somnia",
+        "format": ["cjs"],
+        "bundle": true,
+        "main": "apps/somnia/src/main.ts",
+        "tsConfig": "apps/somnia/tsconfig.app.json",
+        "assets": [
+          "apps/somnia/src/assets",
+          "apps/somnia/src/data",
```
Comment (nguyentvan7) [inline apps/somnia/project.json:21]: Do we need all these paths here?

File: apps/somnia/src/commands/archive.ts [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+  StringSelectMenuInteraction,
+} from 'discord.js'
+import { SlashCommandBuilder } from 'discord.js'
+import * as fs from 'fs'
+import * as path from 'path'
+
+import { error } from '../lib/message'
+import { cwd } from '../util'
+
+import {
```
Comment (nguyentvan7) [inline apps/somnia/src/commands/archive.ts:15]: This should be throwing an error for not importing type

File: apps/somnia/src/commands/archive.ts [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+          .setRequired(true)
+      )
+  )
+
+//requiring all the data because imports dont work
+const archivepath = path.join(cwd, '/libs/gi/dm-localization/assets/locales/en')
+//get keys
+const archive: Record<string, any> = {
+  key: {
+    char: require(path.join(archivepath, '/charNames_gen.json')),
```
Comment (nguyentvan7) [inline apps/somnia/src/commands/archive.ts:79]: change to import

File: apps/somnia/src/commands/archive.ts [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+archive['key']['char']['TravelerAnemoF'] = 'Traveler (Anemo)'
+archive['key']['char']['TravelerGeoF'] = 'Traveler (Geo)'
+archive['key']['char']['TravelerElectroF'] = 'Traveler (Electro)'
+archive['key']['char']['TravelerDendroF'] = 'Traveler (Dendro)'
+archive['key']['char']['TravelerHydroF'] = 'Traveler (Hydro)'
+//get all the data from keys
+for (const category in archive['key']) {
+  for (const name in archive['key'][category]) {
+    const itempath = path.join(archivepath, `/${category}_${name}_gen.json`)
+    if (fs.existsSync(itempath)) archive[category][name] = require(itempath)
```
Comment (nguyentvan7) [inline apps/somnia/src/commands/archive.ts:100]: Is there a way to do this with imports? not sure

File: apps/somnia/src/commands/archive.ts [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+  }
+}
+
+const colors = {
```
Comment (nguyentvan7) [inline apps/somnia/src/commands/archive.ts:104]: These colors and images should exist somewhere in our libraries. Tho they might be SVG, not sure if embeds support that

File: apps/somnia/src/commands/archive.ts [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+  s = s.replaceAll(/(<\/?i>)+/g, '*')
+  //turn rest into bold
+  s = s.replaceAll(/(<\/?\w+>)+/g, '**')
+  //ignore <br/> tags
+  s = s.replaceAll(/<\w+\/>/g, '')
+  //remove extra whitespace
+  return s.trim()
+}
+
+export { archive, clean, colors, talentlist }
```
Comment (nguyentvan7) [inline apps/somnia/src/commands/archive.ts:163]: Just export these when you are creating them (unless this does something different im not aware of)

File: apps/somnia/src/commands/archive.ts [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+  const name = archive['key'][subcommand][id]
+  const data = archive[subcommand][id]
+
+  //handle invalid names
+  if (!(id in archive[subcommand])) throw `Invalid ${subcommand} name.`
+  //character archive
+  if (subcommand === 'char') {
+    return chararchive(
+      interaction,
+      id as LocationCharacterKey,
```
Comment (nguyentvan7) [inline apps/somnia/src/commands/archive.ts:217]: Not sure if we need a type narrowing here. probably fine to leave

File: apps/somnia/src/commands/archive/char.ts [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+export function chararchive(
+  interaction: Interaction,
+  id: LocationCharacterKey,
+  name: string,
+  data: any,
+  args: string
+) {
+  //get element
+  let element = 'none'
+  const travelerelement = id.match(/Traveler(.*)F/)
```
Comment (nguyentvan7) [inline apps/somnia/src/commands/archive/char.ts:24]: getCharEle function from gi-stats

File: apps/somnia/src/commands/archive/char.ts [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+  if (!travelerelement) element = allStats.char.data[id].ele!
+  else element = travelerelement[1].toLowerCase()
+  //setup embed
+  const embed = new EmbedBuilder()
+    .setFooter({
+      text: 'Character Archive',
+    })
+    .setAuthor({
+      name: name,
+      iconURL: colors.element[element as ElementKey | 'none'].img,
```
Comment (nguyentvan7) [inline apps/somnia/src/commands/archive/char.ts:34]: using library should fix the need for type assertion here and below

File: apps/somnia/src/commands/archive/char.ts [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+    .setColor(colors.element[element as ElementKey | 'none'].color)
+  //set contents
+  let text = ''
+  const talent = args
+
+  //character profile
+  if (talent === 'p') {
+    text =
+      data.description ??
+      'A traveler from another world who had their only kin taken away, forcing them to embark on a journey to find The Seven.'
```
Comment (nguyentvan7) [inline apps/somnia/src/commands/archive/char.ts:45]: Is there a better way to do this?

File: apps/somnia/src/commands/archive/char.ts [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+      iconURL: colors.element[element as ElementKey | 'none'].img,
+    })
+    .setColor(colors.element[element as ElementKey | 'none'].color)
+  //set contents
+  let text = ''
+  const talent = args
+
+  //character profile
+  if (talent === 'p') {
+    text =
```
Comment (nguyentvan7) [inline apps/somnia/src/commands/archive/char.ts:43]: For each of these `if (talent ==...)` blocks, extract the body of the if-statement to a function like `text = passiveText(); embed= passiveEmbed()` or something

File: apps/somnia/src/commands/archive/weapon.ts [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+    //r1 for 5 star weapons
+    if (rarity > 4) refine = '1'
+    //r5 for 3/4 star weapons
+    else refine = '5'
+    //user input override
+    if (args) refine = args
+    name += ` (R${refine})`
+    //create dropdown menu
+    const options = []
+    for (let i = 1; i <= 5; i++) {
```
Comment (nguyentvan7) [inline apps/somnia/src/commands/archive/weapon.ts:37]: for future reference, can use `range` util for this `range(1, 5).forEach(i => )`

**Review by nguyentvan7 — CHANGES_REQUESTED** [review]

File: apps/somnia/project.json (hunk 1 above)
Comment (Godlander) [reply apps/somnia/project.json:21]: i expect to use assets in the future for any somnia specific assets that arn't tied to other GO projects. data was meant for storing user data, can be used for temporarily keeping it before that backend database

File: apps/somnia/src/commands/archive.ts (hunk 4 above)
Comment (Godlander) [reply apps/somnia/src/commands/archive.ts:100]: the only reason i load the keys first is to get all the names for that list. i suppose i could hardcode a list of names to import all of them and then export again as one object. maybe it can be generated along with the dm?

File: apps/somnia/src/commands/archive.ts (hunk 6 above)
Comment (Godlander) [reply apps/somnia/src/commands/archive.ts:163]: specifically archive i can't export as i create it, since it gets changed after

File: apps/somnia/src/commands/archive.ts (hunk 5 above)
Comment (Godlander) [reply apps/somnia/src/commands/archive.ts:104]: embeds don't actually take files, only links. so i need an image link

File: apps/somnia/src/commands/archive/char.ts [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+  interaction: Interaction,
+  id: CharacterKey,
+  name: string,
+  data: any,
+  args: string
+) {
+  //get element
+  const element = getCharEle(id)
+  //setup embed
+  let embed = new EmbedBuilder()
```
Comment (nguyentvan7) [inline apps/somnia/src/commands/archive/char.ts:106]: Why do we need a default embed?

You can also wrap all those if-statements below into a function `getEmbed` or something then do 
`const embed = getEmbed()`

File: apps/somnia/src/commands/archive/char.ts [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+      name: name,
+      iconURL: colors.element[element].img,
+    })
+    .setColor(colors.element[element].color)
+  //set contents
+  const talent = args
+
+  //character profile
+  if (talent === 'p') {
+    embed = profileEmbed(embed, id, data)
```
Comment (nguyentvan7) [inline apps/somnia/src/commands/archive/char.ts:120]: Seems like `id` is unused in all these functions

File: apps/somnia/src/commands/archive.ts [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+          .setDescription('Artifact set name')
+          .setAutocomplete(true)
+          .setRequired(true)
+      )
+  )
+
+//requiring all the data because imports dont work
+const archivepath = path.join(cwd, '/libs/gi/dm-localization/assets/locales/en')
+//get keys
+const archive: Record<string, any> = {
```
Comment (nguyentvan7) [inline apps/somnia/src/commands/archive.ts:82]: should fix this typing, but that can be another PR

Comment (nguyentvan7) [reply apps/somnia/src/commands/archive.ts:82]: Can you add a TODO to fix the typing
`// TODO: Fix any typing`

File: apps/somnia/src/commands/archive.ts [hunk 16]
```diff
@@ -0,0 +1,263 @@
...
+      .addStringOption((o) =>
+        o
+          .setName('name')
+          .setDescription('Artifact set name')
+          .setAutocomplete(true)
+          .setRequired(true)
+      )
+  )
+
+//requiring all the data because imports dont work
```
Comment (nguyentvan7) [inline apps/somnia/src/commands/archive.ts:79]: Change this to a TODO:

**Review by nguyentvan7 — CHANGES_REQUESTED** [review]
Just some lint issues, after we should be fine to merge and just deal with everything else as TODO

If you have other things that need fixing, please mark them with `TODO:` as well

File: apps/somnia/src/commands/archive/char.ts (hunk 10 above)
Comment (nguyentvan7) [reply apps/somnia/src/commands/archive/char.ts]: leave a todo to get this from our actual strings

File: apps/somnia/src/commands/archive.ts (hunk 5 above)
Comment (nguyentvan7) [reply apps/somnia/src/commands/archive.ts]: discussed on discord, should be fine to use ambr api for images

File: apps/somnia/src/commands/archive.ts (hunk 4 above)
Comment (nguyentvan7) [reply apps/somnia/src/commands/archive.ts:108]: Leave a todo to clean up this logic

File: apps/somnia/src/commands/archive.ts (hunk 3 above)
Comment (nguyentvan7) [reply apps/somnia/src/commands/archive.ts:84]: leave a todo to clean this up once translation is properly exported by our libraries

**Review by nguyentvan7 — APPROVED** [review]

**Review by nguyentvan7 — APPROVED** [review]
